datalake: allow for translation from tiered storage

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
The high level problem is that without this change, when the Iceberg catalog is down, translation will be blocked and the translation STM's max collectible offset will block local data from being reclaimed by space management. This PR decouples the translation STM's max collectible offset from translation progress by introducing a new concept to the partition, `kafka::offset lowest_pinned_offset()`. Data including and above this offset will not be eligible for compaction or tiered storage retention, and translation progress will move this pinned offset rather than the max collectible offset.

In introducing this concept, the max collectible offset for the translation STM is now only blocked if we have never replicated a highest translated offset. The first replication done by translators is updated to replicate the local start offset - 1 (which may be -1, and is differentiated from `kafka::offset::min()`), ensuring that we retain the semantics of starting translation from local data when translation is first enabled.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Iceberg topics will no longer prevent space management from reclaiming local disk space when data is not translated. Instead, the data to translate will come from tiered storage.
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
cloud_storage: add method for estimating size

Adds a method to estimate the size above a given Kafka offset. My intent
is to make this visible to the datalake subsystem, which will allow it
to estimate how much data needs to be translated for a read replica or
restored topic.

andrwng committed:
kafka: add size estimation to replicated_partition

Adds a size estimation function to the kafka::replicated_partition. I
intend on using this for backlog computation when Iceberg translation is
behind. An upcoming change will allow Iceberg translation to read from
tiered storage, and so the existing local-only backlog computation is
not sufficient.

andrwng committed:
rptest/services: separate out Iceberg REST port

This will be useful to block a specific port.

andrwng committed:
cloud_storage: manifest view Kafka offset pin

Adds a mechanism to pin the spillover retention computation based on a
given Kafka offset. Once Iceberg is able to read from tiered storage,
this will allow us to preserve the semantics that regardless of
retention policy, Kafka records are guaranteed land in Iceberg.

andrwng committed:
archival: archival retention calculator Kafka offset pin

Similar to the pin for spillover retention, this adds a mechanism to pin
archival STM retention based on on a given Kafka offset. Once Iceberg is
able to read from tiered storage, this will allow us to preserve the
semantics that regardless of retention policy, Kafka records are
guaranteed to land in Iceberg.

andrwng commented:
/ci-repeat

Lazin commented on src/v/cloud_storage/async_manifest_view.cc:
```diff
@@ -946,7 +946,8 @@ ss::future<
   std::optional<size_t> size_limit,
-  std::optional<std::chrono::milliseconds> time_limit) noexcept {
+  std::optional<std::chrono::milliseconds> time_limit,
+  std::optional<kafka::offset> pinned_offset) noexcept {
```
This change makes retention calculation non-idempotent because it depends on the parameter now. 

Few questions:
- if the data is about to be deleted by the retention, is it really necessary to translate it and send to iceberg?
- wouldn't it be sufficient to just pause cloud retention while translation is trying to keep up?

andrwng replied:
>if the data is about to be deleted by the retention, is it really necessary to translate it and send to iceberg?

The issue is that users don't really have much control over how fast or slow translation is, or when it gets scheduled. Given that, it's hard to claim any kind of delivery guarantees if we allow data to be removed before translating

>wouldn't it be sufficient to just pause cloud retention while translation is trying to keep up?

Yea that's true. Not a bad idea if trying to get this in quick. I may consider doing this, just to avoid polluting the retention code for now. Though it's certainly a hack -- long term, I think we'll need to update retention to acknowledge the pinned offset.

andrwng replied:
I went the route of just adding unit tests for the relevant subsystems. I'm open to more discussion towards your approach if you think this is still too complex.

I also removed the change from the compaction code and instead relied on the existing compaction config parameter.

Lazin commented on src/v/cloud_storage/async_manifest_view.cc:
```diff
@@ -1005,15 +1007,31 @@ async_manifest_view::compute_retention(
           "Found offset {} to advance start offset to",
           result.offset);
     }
+    if (pinned_offset && result.offset - result.delta > *pinned_offset) {
```
It looks like this computation could be performed outside of this method.

andrwng replied:
Hm I didn't have a clear sense of what you were thinking -- mind clarifying?

Lazin commented on src/v/cluster/archival/retention_calculator.cc:
```diff
@@ -175,19 +176,25 @@ std::optional<retention_calculator> retention_calculator::factory(
     auto it = _manifest.first_addressable_segment();
     for (; it != _manifest.end(); ++it) {
         const auto& entry = *it;
+        if (_pinned_offset && entry.last_kafka_offset() >= *_pinned_offset) {
```
ditto
two subsystems are tied together via parameter and the behaviour is more complicated and nuanced

andrwng replied:
Same here, can you clarify? Not sure how this would work

Lazin reviewed (dismissed):
I didn't finish the review but so far it looks like a lot of complexity is being added. This offset pinning is a viral thing that goes through several subsystems so they depend on each other more (e.g. we have to test TS retention with and without this pinned offset set, same is true for compaction, etc). I'm wondering if there is a way to simplify this somehow.

mmaslankaprv commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -174,13 +174,29 @@ model::offset translation_stm::max_collectible_offset() {
 
-    return highest_log_offset_below_next(
-      _raft->log(), _highest_translated_offset);
+std::optional<kafka::offset> translation_stm::lowest_pinned_offset() const {
```
Does it have to be Kafka offset ?, we are mixing model and Kafka offsets in the same interface

andrwng replied:
For posterity, we chatted about this a bit over a call and while I agree that it's a bit rough to be mixing Kafka and model offsets in the STM manager,
1. we do need this to be Kafka offsets, because the translation STM is unaware of model offsets, especially when reading data from tiered storage
2. _something_ needs to know about both Kafka offsets and model offsets, whether it's the translator STM, or cloud storage readers, or the STM manager. The STM manager just so happens to be the lowest hanging fruit.

I'm open to moving this around. The benefit of the current approach is that there are roughly no on-disk changes, so it seems like it should be easy to move concepts around.

mmaslankaprv commented on src/v/storage/types.cc:
```diff
@@ -31,6 +31,18 @@ model::offset stm_manager::max_collectible_offset() {
+    std::optional<kafka::offset> result;
+    for (const auto& stm : _stms) {
+        auto pinned = stm->lowest_pinned_offset();
+        vlog(stlog.trace, "lowest_pinned_offset[{}] = {}", stm->name(), pinned);
```
i am wondering if this should use storage gc logger ?

andrwng replied:
Removed this log line

andrwng committed:
datalake: pin offsets when not translated

The high level problem is that without this change, when the Iceberg
catalog is down, translation will be blocked and the translation STM's
max collectible offset will block local data from being reclaimed by
space management.

This commit decouples the translation STM's max collectible offset from
translation progress by introducing a new concept to the partition,
`kafka::offset lowest_pinned_data_offset()`. Data including and above
this offset will not be eligible for compaction or tiered storage
retention, and translation progress will move this pinned offset rather
than the max collectible offset.

In introducing this concept, the max collectible offset for the
translation STM is now only blocked if we have never replicated a
highest translated offset. The first replication done by translators is
updated to replicate the local start offset - 1 (which may be -1, and is
differentiated from kafka::offset::min()), ensuring that we retain the
semantics of starting translation from local data when translation is
first enabled.

andrwng commented:
Force push:
- most of the new code is tests
- some tweaks to log lines, small tweaks to the impl, but the approach is the same as the original

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--eeca3b7c-eb72-45a1-a52d-2b5bbe712446-->
<!--TEST RESULTS-->
<!--020b7dc0-c811-4d58-9561-7570739c3acc-->
<!--TEST RESULTS-->
<!--938af312-9ba6-4399-bcae-46e7e05a1995-->
<!--TEST RESULTS-->
<!--a2a2e87f-7f98-4ae9-9e22-47da28e31834-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64451">build#64451</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/64451#019628bb-4bdf-4d86-82c5-05f2d4765e0e | FLAKY | 1/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64451#019628bb-4be0-467f-b59a-107523eb5114 | FLAKY | 1/2 |
| rptest.tests.datalake.blocked_catalog_test.DatalakeBlockedCatalogTest.test_block_cloud_retention_after_translation.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64451#01962915-6119-4bca-b2e1-845872770cb1 | FLAKY | 1/2 |
| rptest.tests.follower_fetching_test.FollowerFetchingTest.test_with_leadership_transfers | ducktape | https://buildkite.com/redpanda/redpanda/builds/64451#01962915-611b-4984-a89f-bb6a6bc24686 | FLAKY | 1/2 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64451#01962915-611a-4d8a-a0b1-1c6a95a69958 | FLAKY | 1/2 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/64451#01962915-611a-4d8a-a0b1-1c6a95a69958 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64524">build#64524</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64524#0196371d-c134-4be2-8a11-3da2b28ccc85 | FLAKY | 1/3 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64524#0196371b-43c8-40bb-b5e6-5ab1027a116b | FLAKY | 1/4 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/64524#0196371b-43c8-40bb-b5e6-5ab1027a116b | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64734">build#64734</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| datalake_cloud_rpfixture.datalake_cloud_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64734#019647b1-3b97-4e3b-a60a-cb8fc9a80aab | FLAKY | 1/2 |
| rptest.tests.audit_log_test.AuditLogTestsAppLifecycle.test_app_lifecycle | ducktape | https://buildkite.com/redpanda/redpanda/builds/64734#019647f1-9bb3-48a5-ae24-22d16e50ff6b | FLAKY | 20/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64734#0196480c-dd71-46d8-81b4-e5beab1ef7af | FLAKY | 17/21 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64734#019647f1-9bb4-439b-8dbf-10f63eba5429 | FAIL | 0/21 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=1 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64734#0196480c-dd72-4e11-b493-f8c22daa6de1 | FAIL | 0/21 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64734#019647f1-9bb2-4e83-ab97-af5a4d3470a6 | FAIL | 0/21 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64734#019647f1-9bb3-48b4-af31-0bb73f02ea25 | FAIL | 0/21 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64734#0196480c-dd71-40bf-b9c1-955f73507195 | FAIL | 0/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64769">build#64769</a></summary>


| test_id                                                                                                                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.availability_test.AvailabilityTests.test_recovery_after_catastrophic_failure                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64769#01964a30-ffe6-41ad-81a4-0f6fbb0fea3e | FLAKY         | 20/21    |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64769#01964a30-ffe5-422d-9ace-0c89216d54d3 | FLAKY         | 11/21    |
| rptest.tests.node_restart_probe_test.NodePreRestartProbeTest.pre_restart_probe_test                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64769#01964a30-ffe7-4901-a8e7-fa20fa277b83 | FLAKY         | 20/21    |
</details>
<!--/body-->

mmaslankaprv commented on src/v/storage/types.h:
```diff
@@ -99,6 +99,11 @@ class snapshotable_stm {
     // log eviction attempts to offsets not greater than this.
     virtual model::offset max_collectible_offset() = 0;
 
+    // Lets the STM limit removal or compaction of data. May refer to a Kafka
```
Would it be possible to add a sentence commenting how this is different from `max_collectible_offset` ?

andrwng replied:
Done

mmaslankaprv commented on src/v/datalake/translation/state_machine.h:
```diff
@@ -30,6 +30,7 @@ class translation_stm final : public raft::persisted_stm<> {
 
     ss::future<> do_apply(const model::record_batch&) override;
     model::offset max_collectible_offset() override;
+    std::optional<kafka::offset> lowest_pinned_offset() const override;
```
maybe it would be worth renaming the two methods to reflect the purpose of them better ?

andrwng replied:
Per our slack conversation, went with:
- max_removable_local_log_offset
- lowest_pinned_data_offset

andrwng committed:
storage: s/max_collectible_offset/max_removable_local_log_offset

With the new pinned offset, it's not immediately clear from the name how
the two differ. So this takes a stab by changing max_collectible_offset
to be max_removable_local_log_offset, indicating 1) it is an offset in
the local log, and 2) it is a log offset rather than a data offset.

Along the way, replaces "collectible" terminology with "removable" -- it
was never really clear to me what "collectible" referred to, but
"removable" seems to capture the fact that the records can be removed
(either by retention or by compaction).

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64734-->
<!--5b39f833-5602-402c-9d9b-f6510a32f72f-->
<!--498b9edd-3baf-4afb-976e-3b1909932eea-->
<!--fe8a4fa2-1218-4a80-941f-ec84e94fd5bf-->
<!--b8650938-fe0f-43c8-a00c-28487a100f56-->
<!--da8e44fb-931b-42ed-aab0-cbf6b58426e0-->
<!--81a5f60d-70e9-4e79-964a-5608c792ab73-->
<!--4d8f9b32-7d3b-4716-86a7-8d71afddbf65-->
<!--title-->
## Retry command for [Build#64734](https://buildkite.com/redpanda/redpanda/builds/64734)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/audit_log_test.py::AuditLogTestsAppLifecycle.test_app_lifecycle
tests/rptest/transactions/consumer_offsets_test.py::VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets@{"versions_to_upgrade":2}
tests/rptest/transactions/consumer_offsets_test.py::VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets@{"versions_to_upgrade":3}
tests/rptest/transactions/consumer_offsets_test.py::VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets@{"versions_to_upgrade":1}
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
```
<!--/body-->

andrwng commented:
Latest force push: updated a test for the new max_collectible_offset name

andrwng commented:
Latest force push: rebase to address a logical conflict in introducing a bazel build for retention_strategy_test

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64769-->
<!--01cfb1ed-aa4e-4215-a449-6ba9d592ad79-->
<!--e151e5e8-be32-49a1-806d-41296dc6e288-->
<!--1ee44447-c908-4345-9d6e-4755aa1527f3-->
<!--title-->
## Retry command for [Build#64769](https://buildkite.com/redpanda/redpanda/builds/64769)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/availability_test.py::AvailabilityTests.test_recovery_after_catastrophic_failure
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/node_restart_probe_test.py::NodePreRestartProbeTest.pre_restart_probe_test
```
<!--/body-->

vbotbuildovich commented:
/backport v25.1.x
